src/utils/containers.ts

```diff
@@ -1,22 +1,95 @@
 import { strict as assert } from "assert";
 import { critical_error, M } from "./debugging-and-logging.js";
-import { set_interval, clear_interval, clear_timeout, set_timeout } from "./node.js";
+import { clear_timeout, set_timeout } from "./node.js";
 
-export class SelfClearingSet<T> {
+let interval_timeout: NodeJS.Timeout | null = null;
+let next_interval: number = Infinity;
+let containers: WeakRef<SelfClearingContainer>[] = [];
+
+function sweep_containers() {
+    containers = containers.filter(c => c.deref() !== undefined);
+    set_next_interval();
+    for (const container of containers) {
+        if (container.deref()!.should_run()) {
+            container.deref()!.sweep();
+        }
+    }
+
+    if (containers.length === 0) {
+        // No more containers, stop the interval
+        interval_timeout = null;
+        next_interval = Infinity;
+        return;
+    }
+}
+
+function set_next_interval() {
+    const timeout_setter = () =>
+        set_timeout(sweep_containers, Math.min(...containers.map(c => c.deref()?.next_interval() ?? Infinity)));
+    setImmediate(timeout_setter); // Unroll the stack, so we don't get a stack overflow
```

Comment:
What causes stack overflow to be a concern here? Shouldn't `set_timeout` similarly schedule something in the event loop without recursing?

Reply:
The main reason I had to add this was that the sweep_containers could have access to the stack, and I've had issues in the past with this approach, so I wanted to make sure it doesn't happen in prod
> Though now that I'm re-reading that should have been a `process.nextTick(timeout_setter)`

Reply:
I see, gotcha